File: dinner/apify_client.py
"""Fetch grocery sales via Apify actor or pre-scraped dataset."""

import json
import re
import time
import urllib.error
import urllib.request

from jumbo_scraper import enrich_deals_with_savings

APIFY_BASE = "https://api.apify.com/v2"
ACTOR_ID = "studio-amba~jumbo-scraper"
# Dutch product keywords → English ingredients for recipe matching
DUTCH_TO_INGREDIENT = [
    (r"\bzalm\b", "salmon"),
    (r"\bkip\b", "chicken"),
    (r"\bvarkens\b|\bbeenham\b|\bham\b", "pork"),
    (r"\brund\b|\bgehakt\b|\bbief\b", "beef"),
    (r"\bgarnalen\b|\bscampi\b", "shrimp"),
    (r"\bbroccoli\b", "broccoli"),
    (r"\btomaat", "tomatoes"),
    (r"\bpaprika\b", "bell pepper"),
    (r"\bappel", "apples"),
    (r"\baardbei", "strawberries"),
    (r"\bavocado\b", "avocado"),
    (r"\bui\b|\buien\b", "onion"),
    (r"\bknoflook\b", "garlic"),
    (r"\bpasta\b|\bspaghetti\b|\bpenne\b", "pasta"),
    (r"\brijst\b", "rice"),
    (r"\bkaas\b|\bcheddar\b|\bkwark\b|\byoghurt\b", "cheese"),
    (r"\broom\b|\bslagroom\b", "cream"),
    (r"\bbonen\b|\bkidney\b", "beans"),
    (r"\bmais\b", "corn"),
    (r"\bspinazie\b|\bsla\b", "spinach"),
    (r"\basperge", "asparagus"),
    (r"\bperen\b|\bperzik", "peaches"),
    (r"\bkomkommer\b", "cucumber"),
    (r"\bwortel", "carrots"),
    (r"\bchampignon", "mushrooms"),
    (r"\bpompoen\b", "pumpkin"),
    (r"\bcitroen\b|\blemon\b", "lemon"),
    (r"\bvis\b|\btilapia\b|\bkabeljauw\b", "fish"),
    (r"\btortilla\b", "tortillas"),
    (r"\bwrap\b", "tortillas"),
]


def _extract_item_name(record):
    if not isinstance(record, dict):
        return None
    for key in ("name", "title", "itemName", "productName", "item"):
        value = record.get(key)
        if isinstance(value, str) and value.strip():
            return value.strip()
    return None


def dutch_name_to_ingredients(name):
    """Map a Dutch sale product name to English ingredient keywords."""
    lowered = name.lower()
    found = []

    for pattern, ingredient in DUTCH_TO_INGREDIENT:
        if re.search(pattern, lowered):
            found.append(ingredient)

    if found:
        return list(dict.fromkeys(found))

    # Fall back to the original name for partial matching
    return [name]


def fetch_dataset_sales(token, dataset_id, limit=500, enrich_savings=False):
    if not token:
        raise ValueError("Apify token is not configured")
    if not dataset_id:
        raise ValueError("Apify dataset ID is not configured")

    url = f"{APIFY_BASE}/datasets/{dataset_id}/items?token={token}&limit={limit}"

    request = urllib.request.Request(url, method="GET")
    try:
        with urllib.request.urlopen(request, timeout=60) as response:
            data = json.loads(response.read().decode("utf-8"))
    except urllib.error.HTTPError as error:
        details = error.read().decode("utf-8", errors="replace")
        raise RuntimeError(f"Apify dataset error ({error.code}): {details}") from error
    except urllib.error.URLError as error:
        raise RuntimeError(f"Could not reach Apify: {error.reason}") from error

    if not isinstance(data, list):
        raise RuntimeError("Unexpected Apify dataset response")

    deals = []
    sale_items = []
    seen_ingredients = set()
    seen_names = set()
    supermarket = None

    for record in data:
        name = _extract_item_name(record)
        if not name:
            continue

        name_key = name.lower()
        if name_key in seen_names:
            continue
        seen_names.add(name_key)

        store = record.get("supermarket") or record.get("merchant") or record.get("store")
        if store and not supermarket:
            supermarket = store

        ingredients = dutch_name_to_ingredients(name)
        for ingredient in ingredients:
            key = ingredient.lower()
            if key not in seen_ingredients:
                seen_ingredients.add(key)
                sale_items.append(ingredient)

        deals.append({
            "name": name,
            "ingredients": ingredients,
            "supermarket": store,
            "price": record.get("price_eur") or record.get("price") or record.get("salePrice"),
            "discount": record.get("discount"),
            "url": record.get("url"),
        })

    if enrich_savings and deals:
        deals = enrich_deals_with_savings(deals)

    if not sale_items:
        raise RuntimeError("No sale items found in the Apify dataset")

    return {
        "source": "apify-dataset",
        "datasetId": dataset_id,
        "supermarket": supermarket,
        "saleItems": sale_items,
        "deals": deals,
        "count": len(deals),
        "ingredientCount": len(sale_items),
    }


def run_actor_and_fetch(token, actor_id=ACTOR_ID, input_data=None, enrich_savings=False, timeout=300):
    """Run an Apify actor, wait for it to finish, and return the results."""
    if not token:
        raise ValueError("Apify token is not configured")

    # Start the actor run
    run_url = f"{APIFY_BASE}/acts/{actor_id}/runs?token={token}"
    payload = json.dumps(input_data or {}).encode("utf-8")
    request = urllib.request.Request(
        run_url,
        data=payload,
        headers={"Content-Type": "application/json"},
        method="POST"
    )
    try:
        with urllib.request.urlopen(request, timeout=30) as response:
            run_data = json.loads(response.read().decode("utf-8"))
    except urllib.error.HTTPError as error:
        details = error.read().decode("utf-8", errors="replace")
        raise RuntimeError(f"Apify actor start error ({error.code}): {details}") from error
    except urllib.error.URLError as error:
        raise RuntimeError(f"Could not reach Apify: {error.reason}") from error

    run_id = run_data["data"]["id"]
    dataset_id = run_data["data"]["defaultDatasetId"]
    print(f"Actor run gestart: {run_id}")

    # Poll until the run is finished
    status_url = f"{APIFY_BASE}/actor-runs/{run_id}?token={token}"
    deadline = time.time() + timeout
    while time.time() < deadline:
        try:
            with urllib.request.urlopen(status_url, timeout=15) as response:
                status_data = json.loads(response.read().decode("utf-8"))
        except urllib.error.URLError as error:
            raise RuntimeError(f"Could not reach Apify: {error.reason}") from error

        status = status_data["data"]["status"]
        print(f"Actor status: {status}")
        if status == "SUCCEEDED":
            break
        if status in ("FAILED", "ABORTED", "TIMED-OUT"):
            raise RuntimeError(f"Actor run mislukt met status: {status}")
        time.sleep(5)
    else:
        raise RuntimeError("Actor run time-out na {timeout} seconden")

    return fetch_dataset_sales(token, dataset_id, enrich_savings=enrich_savings)


def fetch_local_sales(token, postal_code=None, dataset_id=None, enrich_savings=False, **_kwargs):
    """Run the Jumbo scraper actor live and return the results."""
    return run_actor_and_fetch(token, actor_id=ACTOR_ID, enrich_savings=enrich_savings)

File: apify-client.js
/** Fetch grocery sales via Apify (Node — used on Vercel). */

const fs = require("fs");
const path = require("path");

const APIFY_BASE = "https://api.apify.com/v2";
const ACTOR_ID = "harvestedge~dutch-supermarkets-all-11";
const VERCEL_FETCH_TIMEOUT_MS = 8000;
const CACHE_FILE = path.join(__dirname, "sales_cache.json");

const DUTCH_TO_INGREDIENT = [
  [/\bzalm\b/, "salmon"],
  [/\bkip\b/, "chicken"],
  [/\bvarkens\b|\bbeenham\b|\bham\b/, "pork"],
  [/\brund\b|\bgehakt\b|\bbief\b/, "beef"],
  [/\bgarnalen\b|\bscampi\b/, "shrimp"],
  [/\bbroccoli\b/, "broccoli"],
  [/\btomaat/, "tomatoes"],
  [/\bpaprika\b/, "bell pepper"],
  [/\bappel/, "apples"],
  [/\bui\b|\buien\b/, "onion"],
  [/\bknoflook\b/, "garlic"],
  [/\bpasta\b|\bspaghetti\b|\bpenne\b/, "pasta"],
  [/\brijst\b/, "rice"],
  [/\bkaas\b|\bkwark\b/, "cheese"],
  [/\broom\b|\bslagroom\b/, "cream"],
  [/\bspinazie\b|\bsla\b/, "spinach"],
  [/\basperge/, "asparagus"],
  [/\bwortel/, "carrots"],
  [/\bchampignon/, "mushrooms"],
  [/\bcitroen\b/, "lemon"],
  [/\btortilla\b|\bwrap\b/, "tortillas"],
];

function extractItemName(record) {
  if (!record || typeof record !== "object") return null;
  for (const key of ["name", "title", "itemName", "productName", "item"]) {
    const value = record[key];
    if (typeof value === "string" && value.trim()) return value.trim();
  }
  return null;
}

function dutchNameToIngredients(name) {
  const lowered = name.toLowerCase();
  const found = [];
  for (const [pattern, ingredient] of DUTCH_TO_INGREDIENT) {
    if (pattern.test(lowered)) found.push(ingredient);
  }
  return found.length ? [...new Set(found)] : [name];
}

function hasDiscount(record) {
  const discount = record?.discount;
  if (discount == null) return false;
  if (typeof discount === "string") return Boolean(discount.trim());
  return Boolean(discount);
}

async function apifyRequest(url, { method = "GET", body, timeoutMs = 15000 } = {}) {
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), timeoutMs);
  try {
    const response = await fetch(url, {
      method,
      signal: controller.signal,
      headers: body ? { "Content-Type": "application/json" } : undefined,
      body: body ? JSON.stringify(body) : undefined,
    });
    const text = await response.text();
    if (!response.ok) {
      throw new Error(`Apify error (${response.status}): ${text.slice(0, 200)}`);
    }
    return text ? JSON.parse(text) : {};
  } finally {
    clearTimeout(timer);
  }
}

function loadBundledCache() {
  try {
    if (!fs.existsSync(CACHE_FILE)) return null;
    const data = JSON.parse(fs.readFileSync(CACHE_FILE, "utf8"));
    if (!data?.saleItems?.length) return null;
    return {
      ...data,
      source: "bundled-cache",
      cached: true,
    };
  } catch {
    return null;
  }
}

async function withTimeout(promise, timeoutMs) {
  return Promise.race([
    promise,
    new Promise((_, reject) => {
      setTimeout(() => reject(new Error(`Apify fetch timed out after ${timeoutMs}ms`)), timeoutMs);
    }),
  ]);
}

async function getLatestSuccessfulRun(token, actorId = ACTOR_ID) {
  const url = `${APIFY_BASE}/acts/${actorId}/runs?token=${token}&status=SUCCEEDED&limit=1&desc=1`;
  const data = await apifyRequest(url);
  const items = data?.data?.items || [];
  return items[0] || null;
}

async function fetchDatasetSales(token, datasetId, { onSaleOnly = false, supermarketFilter = null, limit = 500, timeoutMs = 15000 } = {}) {
  const url = `${APIFY_BASE}/datasets/${datasetId}/items?token=${token}&limit=${limit}`;
  const data = await apifyRequest(url, { timeoutMs });
  if (!Array.isArray(data)) {
    throw new Error("Unexpected Apify dataset response");
  }

  const deals = [];
  const saleItems = [];
  const seenIngredients = new Set();
  const seenNames = new Set();
  let supermarket = null;

  for (const record of data) {
    if (onSaleOnly && !hasDiscount(record)) continue;

    const store = record.supermarket || record.merchant || record.store;
    if (supermarketFilter) {
      const storeText = `${store || ""} ${record.url || ""}`.toLowerCase();
      if (!storeText.includes(supermarketFilter.toLowerCase())) continue;
    }

    const name = extractItemName(record);
    if (!name) continue;

    const nameKey = name.toLowerCase();
    if (seenNames.has(nameKey)) continue;
    seenNames.add(nameKey);

    if (store && !supermarket) supermarket = store;

    const ingredients = dutchNameToIngredients(name);
    for (const ingredient of ingredients) {
      const key = ingredient.toLowerCase();
      if (!seenIngredients.has(key)) {
        seenIngredients.add(key);
        saleItems.push(ingredient);
      }
    }

    const discount = record.discount;
    let savingsText = null;
    if (hasDiscount(record)) {
      savingsText = typeof discount === "number"
        ? `Was €${discount.toFixed(2).replace(".", ",")}`
        : `Was ${discount}`;
    }

    deals.push({
      name,
      ingredients,
      supermarket: store,
      price: record.price_eur ?? record.price ?? record.salePrice,
      discount,
      url: record.url,
      unitSize: record.unit_size,
      savingsText,
      promotionTag: savingsText,
    });
  }

  if (!saleItems.length) {
    throw new Error("No sale items found in the Apify dataset");
  }

  return {
    source: "apify-dataset",
    datasetId,
    supermarket,
    saleItems,
    deals,
    count: deals.length,
    ingredientCount: saleItems.length,
  };
}

async function fetchFromApify(token, { datasetId, actorId = ACTOR_ID } = {}) {
  const timeoutMs = process.env.VERCEL ? VERCEL_FETCH_TIMEOUT_MS : 60000;
  const itemLimit = process.env.VERCEL ? 120 : 500;

  if (datasetId) {
    const result = await fetchDatasetSales(token, datasetId, { limit: itemLimit, timeoutMs });
    result.source = "apify-dataset";
    return result;
  }

  const latestRun = await apifyRequest(
    `${APIFY_BASE}/acts/${actorId}/runs?token=${token}&status=SUCCEEDED&limit=1&desc=1`,
    { timeoutMs: Math.min(timeoutMs, 5000) }
  );
  const run = latestRun?.data?.items?.[0];
  if (!run?.defaultDatasetId) {
    throw new Error("No recent Apify run found");
  }

  const result = await fetchDatasetSales(token, run.defaultDatasetId, {
    onSaleOnly: true,
    supermarketFilter: "jumbo",
    limit: itemLimit,
    timeoutMs,
  });
  result.source = "apify-actor-run";
  result.actorId = actorId;
  result.runId = run.id;
  result.datasetId = run.defaultDatasetId;
  return result;
}

async function fetchLocalSales(token, { datasetId, actorId = ACTOR_ID, refresh = false } = {}) {
  if (!token) throw new Error("Apify token is not configured");

  // Vercel functions time out quickly — serve bundled cache unless user explicitly refreshes.
  if (process.env.VERCEL && !refresh) {
    const cached = loadBundledCache();
    if (cached) return cached;
  }

  try {
    const fetchPromise = fetchFromApify(token, { datasetId, actorId });
    const timeoutMs = process.env.VERCEL ? VERCEL_FETCH_TIMEOUT_MS : 60000;
    return await withTimeout(fetchPromise, timeoutMs);
  } catch (error) {
    const cached = loadBundledCache();
    if (cached) {
      cached.fallbackReason = error.message;
      return cached;
    }
    throw error;
  }
}

module.exports = {
  ACTOR_ID,
  fetchLocalSales,
  loadBundledCache,
  getLatestSuccessfulRun,
};
